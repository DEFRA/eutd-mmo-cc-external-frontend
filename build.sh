#/bin/bash
curl -O https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz
tar -xf node-v12.18.4-linux-x64.tar.xz --directory /tmp
export PATH="/tmp/node-v12.18.4-linux-x64/bin:$PATH"
rm node-v12.18.4-linux-x64.tar.xz
npm ci
npm run test-vstack
rm -rf /tmp/node-v12.18.4-linux-x64