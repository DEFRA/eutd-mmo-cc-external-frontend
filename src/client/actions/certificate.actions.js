import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};