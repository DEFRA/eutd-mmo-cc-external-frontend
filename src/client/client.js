// Startup point for the client side application
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Router, withRouter} from 'react-router-dom';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import axios from 'axios';
import Routes from './Routes';
import reducers from './reducers';
import { createBrowserHistory } from 'history';

const composeEnhancers = process.env.NODE_ENV === 'development' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__  ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose;
  

const paramSerializerFn = params => {
  let output = '';
  Object.keys(params).forEach((k) => {
    output+= `${k}=${params[k]}&`;
  });
  return encodeURI(output);
};

const axiosResponseInterceptorSuccessFn = response => response;
const axiosResponseInterceptorErrorFn = error => {
  if (error.response.status === 599) {
    // Let's check if this is because we're logged out
    const empPromise = new Promise(() => {});
    const searchParam = error.request.responseURL.slice(error.request.responseURL.search(/[?]/g));
    let searchParams = new URLSearchParams(searchParam);
    if(searchParams.get('notLoggedInErr') === 'yes') {
      window.location = '/logout?backToPath=/server-logout';
      return empPromise;
    }
    // It's always a timeout...
    window.location = '/timed-out';
    return new Promise(() => {});
  } else {
    return new Promise((_,rej) => rej(error));
  }
};

const axiosInstanceForOrchService = axios.create({
  baseURL: '/orchestration/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForOrchService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForRefService = axios.create({
  baseURL: '/reference/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForRefService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForDynamixService = axios.create({
  baseURL: '/dynamix/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForDynamixService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const store = createStore(
  reducers,
  window.INITIAL_STATE,
  composeEnhancers(applyMiddleware(
    // The difference between orch/ref service api is not significant as they both go through proxy
    // However it is significant in the server side store. To stick with same action creators we're
    // passing down same axios instance to both orchestration and reference service api .
    thunk.withExtraArgument( {
      orchestrationApi: axiosInstanceForOrchService,
      referenceServiceApi: axiosInstanceForRefService,
      dynamixApi: axiosInstanceForDynamixService } )
  ))
);

const ScrollToTop = withRouter( class ScrollToTop extends React.Component {
  componentDidMount() {
  }

  render() {
    return this.props.children;
  }
});

const history = createBrowserHistory();


ReactDOM.hydrate(
  <Provider store={store}>
    <Router history={history}>
      <ScrollToTop>
        <div>{renderRoutes(Routes)}</div>
      </ScrollToTop>
    </Router>
  </Provider>,
  document.querySelector('#root')
);
