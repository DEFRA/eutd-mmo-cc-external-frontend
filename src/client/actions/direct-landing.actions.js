import { beginApiCall, apiCallFailed, isForbidden, showFullPageError, CLEAR_ERRORS, isSilverLineError } from './index';
import { exportPayloadActionTypes } from '../actions';


export const GET_DIRECT_LANDING_PRODUCTS = 'get_direct_landing_products';
export const CLEAR_DIRECT_LANDING = 'clear_direct_landing';
export const DIRECT_LANDING_PRODUCTS_FAILED = 'direct_landing_products_failed';
export const POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'post_direct_landing_products_unauthorised';
export const GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'get_direct_landing_products_unauthorised';
export const POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR = 'post_direct_landing_prodcts_waf_error';

export const getDirectLandings =
  (documentNumber) =>
  async (dispatch, _getState, { orchestrationApi }) => {
    const config = { headers: { documentnumber: documentNumber } };

    try {
      dispatch(beginApiCall());

      const res = await orchestrationApi.get(
        '/export-certificates/export-payload/direct-landings',
        config
      );

      await dispatch({ type: GET_DIRECT_LANDING_PRODUCTS, payload: res.data });
    } catch (e) {
      if (isForbidden(e)) {
        dispatch({ type: GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
      } else {
        dispatch(apiCallFailed(e.response));
        throw new Error(
          `An error has occurred ${e.response.status} (${e.response.statusText})`
        );
      }
    }
  };

export const clearDirectLanding = () => ({ type: CLEAR_DIRECT_LANDING });

export const upsertDirectLanding = (landing, documentNumber = undefined) => (dispatch, getState, { orchestrationApi } ) => {
  const config = (documentNumber)
    ? {
        headers: {
          documentnumber: documentNumber
        }
    }
    : {};

  return orchestrationApi.post('/export-certificates/direct-landing/validate', landing, config)
    .then(
      res => {
        dispatch(landingUpserted(res.data));
        dispatch(clearErrors());
      },
      error => {
        if (error.response.status === 400) {
          dispatch(landingUpserted(error.response.data));
          throw new Error();
        } else if (isSilverLineError(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR, supportID: error.response.data.match(/\d+/)[0] });
        } else if (isForbidden(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
        } else {
          dispatch(showFullPageError());
        }
      });

  function landingUpserted(exportPayload) { return { type: exportPayloadActionTypes.LANDING_UPSERTED, exportPayload, errors: exportPayload.errors }; }
  function clearErrors() { return { type: CLEAR_ERRORS }; }
};
