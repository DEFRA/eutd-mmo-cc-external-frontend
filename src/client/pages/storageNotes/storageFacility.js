import React from 'react';
import {connect} from 'react-redux';
import _ from 'lodash';
import {withRouter} from 'react-router-dom';
import { withTranslation } from 'react-i18next';

import {
  InputField,
  Header,
  GridRow,
  GridCol,
  Main,
  BackLink,
} from 'govuk-react';

import {
  getStorageNotesFromRedis,
  saveStorageNotes,
  saveStorageNotesToRedis,
  clearStorageNotes,
  changeStorageFacilityAddress
} from '../../actions';
import {clearPostcodeLookupAddress} from '../../actions/postcode-lookup.actions';

import PageTitle from '../../components/PageTitle';
import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import ErrorIsland from '../../components/elements/ErrorIsland';
import {
  onHandleErrorClick,
  scrollToErrorIsland,
  toGovukErrors
} from '../utils';
import Form from '../../components/elements/Form';
import HelpLink from '../../components/HelpLink';
import SaveAsDraftButton from '../../components/SaveAsDraftButton';
import ContinueButton from '../../components/elements/ContinueButton';
import  LookupAddress  from '../../components/LookupAddress';
import NotificationBanner from '../../components/NotificationBanner';
import BackToProgressLink from '../../components/BackToProgressLink';

class StorageFacility extends React.Component {

  constructor(props) {
    super(props);
    this.state = this.init(props.storageNotes, props);
  }

  init(data, props) {

    const storageNotes = _.cloneDeep(data || {});
    const index = +(props.match.params.facilityIndex) || 0;

    if (storageNotes.storageFacilities && index >= storageNotes.storageFacilities.length) {
      storageNotes.storageFacilities.push({});
    }

    const postcodeLookupAddress = (this.props.postcodeLookup && this.props.postcodeLookup.postcodeLookupAddress)
      ? this.props.postcodeLookup.postcodeLookupAddress
      : {};

    if (!_.isEmpty(postcodeLookupAddress)) {
      storageNotes.storageFacilities[index] = {
        ...storageNotes.storageFacilities[index],
        ...postcodeLookupAddress
      };
    }

    return ({
      index,
      storageNotes,
      errors: storageNotes.errors || {},
      errorsUrl: storageNotes.errorsUrl,
      species: ''
    });

  }

  goBack(e, page) {
    e.preventDefault();
    const storageNotesClone = _.cloneDeep(this.props.storageNotes || {});
    storageNotesClone.errors = {};
    this.props.save(storageNotesClone);
    this.props.history.push(page);
  }

  onChange(e, name) {
    const {index, storageNotes} = this.state;
    if(storageNotes && storageNotes.storageFacilities[index]){
      storageNotes.storageFacilities[index][name] = e.target.value;
      this.props.save(storageNotes);
      this.setState({unsavedFacilityName: e.target.value});
      this.setState({storageNotes});
    }
  }

  async onContinue(nextUri) {
    const {index} = this.state;
    const {documentNumber} = this.props.match.params;

    const currentUrl = this.props.route.firstStorageFacility ?
      `/create-storage-document/${documentNumber}/add-storage-facility-details`
      : `/create-storage-document/${documentNumber}/add-storage-facility-details/${index}`;

    const {saveAsDraftUri} = this.props.route;
    const isStorageFacilitySavedAsDraft = nextUri === saveAsDraftUri ? true : false;
    const storageNotes = await this.props.saveToRedis({
      saveToRedisIfErrors: isStorageFacilitySavedAsDraft,
      data: this.state.storageNotes,
      currentUrl: currentUrl,
      documentNumber: documentNumber,
    });

    if (!isStorageFacilitySavedAsDraft && !_.isEmpty(storageNotes.errors)) {
      this.setState({
        errors: storageNotes.errors,
        errorsUrl: storageNotes.errorsUrl
      });
      scrollToErrorIsland();
      setTimeout(
        () =>
          document
            .getElementById('errorIsland')
            .scrollIntoView({behavior: 'smooth'}),
        0
      );
      return;
    }
    this.props.clearPostcodeLookupAddress();
    this.props.history.push(nextUri.replace(':documentNumber', documentNumber));
  }

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    const {saveAsDraftUri} = this.props.route;
   await this.onContinue(saveAsDraftUri);
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  componentDidUpdate() {
    const { storageNotes } = this.props;

    if (storageNotes.unauthorised === true) {
      this.unauthorised();
    }
  }

  async componentDidMount() {
    window.scrollTo(0, 0);
    const {documentNumber} = this.props.match.params;

    if (!this.props.storageNotes.changeAddress) {
      const data = await this.props.getFromRedis(this.props.match.params.documentNumber);
      this.setState(this.init(data, this.props));
    }
    const {index, storageNotes} = this.state;
    const facility = storageNotes.storageFacilities[index];
    if (facility === undefined) {
      this.props.history.push(`/create-storage-document/${documentNumber}/you-have-added-a-storage-facility`);
    }
  }

  componentWillUnmount() {
    this.props.clear();
  }

  render() {

    const {history, route, match, t} = this.props;

    const {journey, saveAsDraftUri, path, changeAddressUri, progressUri} = route;
    const {documentNumber} = match.params;
    const {errors, errorsUrl, index, storageNotes} = this.state;
    const facility = storageNotes.storageFacilities && storageNotes.storageFacilities[index];

    const currentUri = path.replace(':facilityIndex', index);
    const changeAddressLink = changeAddressUri.replace(':documentNumber', documentNumber).replace(':facilityIndex', parseInt(index));

    if (!facility) return null;
    const {
      facilityName = '',
      facilityAddressOne = '',
      facilityTownCity = '',
      facilityPostcode = '',
    } = facility;

    const errs = (errorsUrl === history.location.pathname && errors) || {}; // only display errors for this URL
    const backLink = this.props.route.firstStorageFacility
      ? `${documentNumber}/you-have-added-a-product`
      : `${documentNumber}/you-have-added-a-storage-facility`;

    const hasFacilityAddressOne = () => storageNotes.changeAddress
      && storageNotes.storageFacilities
      && storageNotes.storageFacilities[index]
      && storageNotes.storageFacilities[index].facilityAddressOne;

    const hasFacilityUpdatedFlag = () => storageNotes.storageFacilities
      && storageNotes.storageFacilities[index]
      && storageNotes.storageFacilities[index]._facilityUpdated;

    const facilitiesUpdated = !hasFacilityAddressOne()
      && hasFacilityUpdatedFlag();

    const saveAsDraftFormActionUrl = `/orchestration/api/v1/storageNotes/saveAndValidate?saveAsDraftUrl=${saveAsDraftUri}&c=/create-storage-document/add-storage-facility-details/${index}&n=/create-storage-document/you-have-added-a-storage-facility`;

    return (
      <Main className="create-storage-document">
        <ErrorIsland
          errors={toGovukErrors(errs)}
          onHandleErrorClick={onHandleErrorClick}
        />
        <PageTitle
          title={`${
            !_.isEmpty(errors) ? 'Error: ' : ''
          }${t('sdAddStorageDetailsHeader')} - ${t('sdCommonTitle')}`}
        />

        <GridRow>
          <GridCol>
            <BackLink
              href={`/orchestration/api/v1/storageNotes/back?n=/create-storage-document/${backLink}`}
              onClick={(e) =>
                this.goBack(e, `/create-storage-document/${backLink}`)
              }
            >
              {t('commonBackLinkBackButtonLabel')}
            </BackLink>
            {facilitiesUpdated && (
              <NotificationBanner
              header={t('commonImportant')}
              messages={[t('sdYouAddedStorageNotificationText')]} />
            )}
            <Header>{t('sdAddStorageDetailsHeader')}</Header>
          </GridCol>
        </GridRow>
        <Form
          action="/orchestration/api/v1/storageNotes/saveAndValidate"
          currentUrl={`/create-storage-document/${documentNumber}/add-storage-facility-details/${index}`}
          nextUrl={`/create-storage-document/${documentNumber}/you-have-added-a-storage-facility`}
          onSubmit={(e) => e.preventDefault()}
        >
          <GridRow>
            <GridCol columnTwoThirds>
              <InputField
                id={`storageFacilities-${index}-facilityName`}
                htmlFor={`storageFacilities.${index}.facilityName`}
                meta={{
                  error: t(errs[`storageFacilities-${index}-facilityName`]),
                  touched: true,
                }}
                input={{
                  autoComplete: 'off',
                  className: 'formControl',
                  id: `storageFacilities.${index}.facilityName`,
                  name: `storageFacilities.${index}.facilityName`,
                  value: _.isEmpty(facilityName) ? storageNotes.unsavedFacilityName : facilityName,
                  onChange: (e) => this.onChange(e, 'facilityName'),
                }}
              >
                {t('sdCommonFacilityNameTitle')}
              </InputField>
            </GridCol>
          </GridRow>
          <GridRow>
            <GridCol columnTwoThirds>
              <LookupAddress
                addressOne={facilityAddressOne}
                townCity={facilityTownCity}
                postcode={facilityPostcode}
                addressType='storage facility'
                changeAddressLink={changeAddressLink}
                changeAddressHandler={this.props.changeStorageFaclityAddress}
                unsavedFacilityName={this.state.unsavedFacilityName}
              />
            </GridCol>
          </GridRow>

          <GridRow>
            <SaveAsDraftButton
              formactionUrl={saveAsDraftFormActionUrl}
              onClick={this.onSaveAsDraft}
            />
            <input type="hidden" name="journey" value={journey} />
            <input type="hidden" name="currentUri" value={currentUri} />
            <ContinueButton
              type="submit"
              id="continue"
              onClick={() =>
                this.onContinue(
                  `/create-storage-document/${documentNumber}/you-have-added-a-storage-facility`
                )
              }
            >
              {t('commonContinueButtonSaveAndContinueButton')}
            </ContinueButton>
          </GridRow>
        </Form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey} />
      </Main>
    );
  }
}

function mapStateToProps(state) {
  return {
    storageNotes: state.storageNotes,
    postcodeLookup: state.postcodeLookup
  };
}

function loadData(store) {
  return store.dispatch(getStorageNotesFromRedis(this.documentNumber));
}

export const component = withRouter(
  connect(mapStateToProps, {
    save: saveStorageNotes,
    saveToRedis: saveStorageNotesToRedis,
    getFromRedis: getStorageNotesFromRedis,
    clear: clearStorageNotes,
    clearPostcodeLookupAddress,
    changeStorageFaclityAddress: changeStorageFacilityAddress
  }
  )(withTranslation()(StorageFacility))
);

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
