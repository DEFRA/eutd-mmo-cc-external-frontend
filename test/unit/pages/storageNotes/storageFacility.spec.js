import { mount } from 'enzyme/build';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { component as StorageFacility } from '../../../../src/client/pages/storageNotes/storageFacility';
import storageFacility from '../../../../src/client/pages/storageNotes/storageFacility';
import { getStorageNotesFromRedis, clearStorageNotes } from '../../../../src/client/actions';
import { render } from '@testing-library/react';

jest.mock('../../../../src/client/actions');

window.scrollTo = jest.fn();

const mockStore = configureStore([thunk.withExtraArgument({
  orchestrationApi: {
    get: () => {
      return new Promise((res) => {
        res({});
      });
    }
  }
})]);

describe('StorageFacility - without storage notes', () => {

  const store = mockStore({
    global: {
      allFish: []
    },
    exporter: {
      model: {
        exporterFullName: 'Mr Fish Exporter',
        exportCompanyName: 'MailOrderFish Ltd',
        addressOne: '5 Costa Road',
        townCity: 'Hebburn',
        postcode: 'NE1 4FS',
        _updated: true
      }
    },
    transport: {
      vehicle: 'truck',
      cmr: 'true',
      exportedTo: {
        isoCodeAlpha2: 'Alpha2',
        isoCodeAlpha3: 'Alpha3',
        isoNumericCode: 'IsoNumericCode',
        officialCountryName: 'Brazil',
      }
    },
    fish: [{}]
    }
  );


  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <StorageFacility route={{ previousUri: '', changeAddressUri: '', journey: 'storageNotes', path: '/products' }} />
      </MemoryRouter>
    </Provider>
  );

  it('should render when no storage facilities are provided', () => {
    expect(wrapper.find('StorageFacility').exists()).toBeTruthy();
    expect(wrapper.find('#continue').exists()).toBeFalsy();
  });
});

describe('StorageFacility - with storage notes and no storage facilities', () => {

  const store = mockStore({
    global: {
      allFish: []
    },
    exporter: {
      model: {
        exporterFullName: 'Mr Fish Exporter',
        exportCompanyName: 'MailOrderFish Ltd',
        addressOne: '5 Costa Road',
        townCity: 'Hebburn',
        postcode: 'NE1 4FS',
        _updated: true
      }
    },
    storageNotes: {
      addAnotherProduct: 'notset',
      addAnotherStorageFacility: 'notset',
      catches: [{
        product: 'Asda squid burgers',
        commodityCode: '123456',
        certificateNumber: '12345',
        productWeight: '1000',
        dateOfUnloading: '25/01/2019',
        placeOfUnloading: 'Dover',
        transportUnloadedFrom: 'Car',
        weightOnCC: '500'
      },
      {
        product: 'Tesco squid burgers',
        commodityCode: '123457',
        certificateNumber: '12346',
        productWeight: '1001',
        dateOfUnloading: '25/01/2019',
        placeOfUnloading: 'Dover',
        transportUnloadedFrom: 'Truck'
      }],
      validationErrors: [{}],
      storageFacilities: []
    },
    transport: {
      vehicle: 'truck',
      cmr: 'true',
      exportedTo: {
        isoCodeAlpha2: 'Alpha2',
        isoCodeAlpha3: 'Alpha3',
        isoNumericCode: 'IsoNumericCode',
        officialCountryName: 'Brazil',
      }
    },
    fish: [{}]
    }
  );

  const props = {
    route: {
      previousUri: '',
      progressUri: '/create-storage-document/progress',
      changeAddressUri: '',
      journey: 'storageNotes',
      path: '/products',
      firstStorageFacility: true
    },
    match: {
      params: ''
    },
    clear: jest.fn()
  };

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <StorageFacility {...props} />
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    clearStorageNotes.mockReturnValue({ type: 'CLEAR_STORAGE_NOTES' });
  });

  it('should render when no storage facilities are provided', () => {
    expect(wrapper.find('StorageFacility').exists()).toBeTruthy();
    expect(wrapper.find('#continue').exists()).toBeTruthy();
  });

  it('should have a back to the progress page link', () => {
    expect(wrapper.find('a[href="/create-storage-document/progress"]').exists()).toBeTruthy();
  });


  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
});

describe('StorageFacility - with storage notes and storage facilities', () => {

  const state = {
    global: {
      allFish: []
    },
    exporter: {
      model: {
        exporterFullName: 'Mr Fish Exporter',
        exportCompanyName: 'MailOrderFish Ltd',
        addressOne: '5 Costa Road',
        townCity: 'Hebburn',
        postcode: 'NE1 4FS',
        _updated: true
      }
    },
    storageNotes: {
      addAnotherProduct: 'notset',
      addAnotherStorageFacility: 'notset',
      catches: [{
        product: 'Asda squid burgers',
        commodityCode: '123456',
        certificateNumber: '12345',
        productWeight: '1000',
        dateOfUnloading: '25/01/2019',
        placeOfUnloading: 'Dover',
        transportUnloadedFrom: 'Car',
        weightOnCC: '500'
      },
      {
        product: 'Tesco squid burgers',
        commodityCode: '123457',
        certificateNumber: '12346',
        productWeight: '1001',
        dateOfUnloading: '25/01/2019',
        placeOfUnloading: 'Dover',
        transportUnloadedFrom: 'Truck'
      }],
      validationErrors: [{}],
      storageFacilities: [
        {
          facilityName: 'FACILITY 1',
          facilityAddressOne: 'BUILDING 1, STREET 1',
          facilityBuildingName: 'BUILDING 1',
          facilityStreetName: 'STREET 1',
          facilityTownCity: 'TOWN 1',
          facilityCounty: 'COUNTY 1',
          facilityCountry: 'COUNTRY 1',
          facilityPostcode: 'POSTCODE 1'
        }
      ]
    },
    transport: {
      vehicle: 'truck',
      cmr: 'true',
      exportedTo: {
        isoCodeAlpha2: 'Alpha2',
        isoCodeAlpha3: 'Alpha3',
        isoNumericCode: 'IsoNumericCode',
        officialCountryName: 'Brazil',
      }
    },
    fish: [{}],
    index: 0
  };

  const props = {
    route: {
      previousUri: '',
      progressUri: '/create-storage-document/progress',
      changeAddressUri: '',
      journey: 'storageNotes',
      path: '/products',
      firstStorageFacility: true
    }
  };

  const store = mockStore(state);

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <StorageFacility {...props} />
      </MemoryRouter>
    </Provider>
  );

  it('will render when storage facilities are provided', () => {
    expect(wrapper.exists('StorageFacility')).toBe(true);
    expect(wrapper.exists('#continue')).toBe(true);
  });

  describe('when populating from state', () => {

    it('will populate the facility name', () => {
      expect(wrapper.exists('[id=\'storageFacilities.0.facilityName\']')).toBe(true);

      expect(wrapper.find('[id=\'storageFacilities.0.facilityName\']').at(0).props()['value']).toBe('FACILITY 1');
    });

    it('will populate the facility address from state', () => {
      expect(wrapper.exists('LookupAddress')).toBe(true);

      expect(wrapper.find('LookupAddress').at(0).props()['addressOne']).toBe('BUILDING 1, STREET 1');
      expect(wrapper.find('LookupAddress').at(0).props()['townCity']).toBe('TOWN 1');
      expect(wrapper.find('LookupAddress').at(0).props()['postcode']).toBe('POSTCODE 1');
    });

  });

  describe('when populating from postcode lookup', () => {

    const stateWithPostcodeLookup = {
      ...state,
      postcodeLookup: {
        postcodeLookupAddress: {
          facilityAddressOne: 'BUILDING 2, STREET 2',
          facilityBuildingName: 'BUILDING 2',
          facilityStreetName: 'STREET 2',
          facilityTownCity: 'TOWN 2',
          facilityCounty: 'COUNTY 2',
          facilityCountry: 'COUNTRY 2',
          facilityPostcode: 'POSTCODE 2'
        }
      }
    };

    const props = {
      route: {
        previousUri: '',
        progressUri: '/create-storage-document/progress',
        changeAddressUri: '',
        journey: 'storageNotes',
        path: '/products',
        firstStorageFacility: true
      }
    };

    const storeWithPostcodeLookup = mockStore(stateWithPostcodeLookup);

    const wrapperWithPostcodeLookup = mount(
      <Provider store={storeWithPostcodeLookup}>
        <MemoryRouter>
          <StorageFacility {...props} />
        </MemoryRouter>
      </Provider>
    );

    it('will override the address in state', () => {
      expect(wrapperWithPostcodeLookup.exists('LookupAddress')).toBe(true);

      expect(wrapperWithPostcodeLookup.find('LookupAddress').at(0).props()['addressOne']).toBe('BUILDING 2, STREET 2');
      expect(wrapperWithPostcodeLookup.find('LookupAddress').at(0).props()['townCity']).toBe('TOWN 2');
      expect(wrapperWithPostcodeLookup.find('LookupAddress').at(0).props()['postcode']).toBe('POSTCODE 2');
    });

    it('will not override the facility name', () => {
      expect(wrapperWithPostcodeLookup.exists('[id=\'storageFacilities.0.facilityName\']')).toBe(true);

      expect(wrapperWithPostcodeLookup.find('[id=\'storageFacilities.0.facilityName\']').at(0).props()['value']).toBe('FACILITY 1');
    });

  });

});

describe('loadData', () => {

  const store = {
    dispatch: () => {
      return new Promise((res) => {
        res();
      });
    }
  };

  const journey = 'storageDocument';
  const documentNumber = 'some-document-number';

  beforeEach(() => {
    getStorageNotesFromRedis.mockReturnValue({ type: 'GET_STORAGE_NOTES_FROM_REDIS' });
  });

  it('will call all methods needed to load the component', async () => {
    storageFacility.documentNumber = documentNumber;

    await storageFacility.loadData(store, journey);

    expect(getStorageNotesFromRedis).toHaveBeenCalledWith(documentNumber);
  });

});

describe('StorageFacility - unauthorised', () => {
  const mockPush = jest.fn();
  const mockClear = jest.fn();

  const props = {
    match: {
      params: {
        documentNumber: 'some-document-number'
      }
    },
    history: {
      push: mockPush,
    },
    storageNotes: {
      unauthorised: true,
      changeAddress: true,
      storageFacilities: []
    },
    index: 0,
    getFromRedis: jest.fn(),
    clear: mockClear
  };

  it('should push history for componentDidUpdate with /forbidden when unauthorised', () => {
    new StorageFacility.WrappedComponent(props).componentDidUpdate();
    expect(mockPush).toHaveBeenCalledWith('/forbidden');
  });

  it('should clear the storage document', () => {
    new StorageFacility.WrappedComponent(props).componentWillUnmount();
    expect(mockClear).toHaveBeenCalledTimes(1);
  });
});
