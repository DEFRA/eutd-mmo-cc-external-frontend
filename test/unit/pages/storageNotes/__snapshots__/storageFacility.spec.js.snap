// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageFacility - with storage notes and no storage facilities should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="create-storage-document css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="/orchestration/api/v1/storageNotes/back?n=/create-storage-document/undefined/you-have-added-a-product"
          >
            Back
          </a>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            Add storage facility details
          </h1>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/storageNotes/saveAndValidate?c=/create-storage-document/undefined/add-storage-facility-details/0&n=/create-storage-document/undefined/you-have-added-a-storage-facility"
        method="POST"
      >
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <label
              class="e1vcssft0 e1vcssft0 css-5tg97n-StyledLabel-StyledHoc-StyledHoc e4u1u2g0"
              for="storageFacilities.0.facilityName"
              id="storageFacilities-0-facilityName"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Facility name
              </span>
              
              <input
                autocomplete="off"
                class="formControl e1vcssft0 css-4n7f6s-StyledInput-StyledHoc e4u1u2g0"
                id="storageFacilities.0.facilityName"
                name="storageFacilities.0.facilityName"
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <p
              id="lookup-address"
            >
              An address must be added for this storage facility.
            </p>
            <p
              id="address-link-wrapper"
            >
              <a
                class="change-address-link"
                href="/"
                id="add-lookup-address-link"
              >
                Add the storage facility address
              </a>
            </p>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/storageNotes/saveAndValidate?saveAsDraftUrl=undefined&c=/create-storage-document/add-storage-facility-details/0&n=/create-storage-document/you-have-added-a-storage-facility"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <input
            name="journey"
            type="hidden"
            value="storageNotes"
          />
          <input
            name="currentUri"
            type="hidden"
            value="/products"
          />
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
            type="submit"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/create-storage-document/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
